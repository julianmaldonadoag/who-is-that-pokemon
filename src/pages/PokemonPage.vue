<template>
  <h1 v-if="!pokemon">Espero un momento por favor...</h1>
  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <PokemonPicture 
      :pokemon-id="pokemon.id" 
      :show-pokemon="showPokemon" 
    />
    <PokemonOptions 
      :pokemons="pokemonArr" 
      @chosen-pokemon="checkAnswer" 
    />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        Nuevo juego
      </button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture' 
import PokemonOptions from '@/components/PokemonOptions' 
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
  components: {
    PokemonPicture,
    PokemonOptions
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions()

      const randomIndex = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[randomIndex]
    },
    checkAnswer(pokemonId) {
      this.showPokemon = true
      this.showAnswer = true

      if (pokemonId === this.pokemon.id) {
        this.message = `Correcto, ${this.pokemon.name}`
      } else {
        this.message = `Oops, era ${this.pokemon.name}`
      }
    },
    newGame() {
      this.pokemonArr = []
      this.pokemon = null
      this.showPokemon = false
      this.showAnswer = false
      this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }
}
</script>
