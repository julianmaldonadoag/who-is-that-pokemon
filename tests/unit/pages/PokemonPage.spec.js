import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage.vue"
import { pokemons } from "../mocks/pokemons.mock"

describe('PokemonPage component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test('should match with snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should call mixPokemonArray when component is mounted', () => {
    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
    const wrapper = shallowMount(PokemonPage)
    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })

  test('should match with snapshot when pokemons are loaded', () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show PokemonPicture and PokemonOptions components', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    const picture = wrapper.find('pokemon-picture-stub')
    const options = wrapper.find('pokemon-options-stub')

    expect(picture.exists()).toBeTruthy()
    expect(options.exists()).toBeTruthy()

    expect(picture.attributes('pokemonid')).toBe('1')
    expect(options.attributes('pokemons')).toBeTruthy()
  })

  test('checkAnswer tests', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    await wrapper.vm.checkAnswer(1) // Correct pokemon

    expect(wrapper.find('h2').exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.vm.message).toBe(`Correcto, ${pokemons[0].name}`)

    await wrapper.vm.checkAnswer(4) // Incorrect pokemon

    expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`)

  })
})