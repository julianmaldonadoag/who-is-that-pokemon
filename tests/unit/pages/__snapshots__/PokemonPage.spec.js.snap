// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match with snapshot 1`] = `
<div>
  <h1>Espere un momento por favor...</h1><img src="" alt="">
</div>
`;

exports[`PokemonPage component should match with snapshot when pokemons are loaded 1`] = `
<div>
  <h1>¿Quién es ese pokémon?</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="pokemon" class="hidden-pokemon"></div>
  <div class="options-container">
    <ul>
      <li class="">bulbasaur</li>
      <li class="">ivysaur</li>
      <li class="">venusaur</li>
      <li class="">charmander</li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;
